diff -up ipmitool-1.8.11/lib/ipmi_main.c.re ipmitool-1.8.11/lib/ipmi_main.c
--- ipmitool-1.8.11/lib/ipmi_main.c.re	2014-09-15 16:17:48.384997935 +0200
+++ ipmitool-1.8.11/lib/ipmi_main.c	2014-09-15 16:28:59.526014566 +0200
@@ -325,28 +325,13 @@ ipmi_main(int argc, char ** argv,
 	char sol_escape_char = SOL_ESCAPE_CHARACTER_DEFAULT;
 	char * devfile  = NULL;
 
-	int cnt = argc;                                                              
-	char **arg = argv;                                                           
-	int voptind;                                                                 
-	int argecnt = 0;                                                             
-	char **arge = NULL;                                                          
-	char **narge = NULL;                                                         
-	char *argestr = NULL;                                                        
-
 	/* save program name */
 	progname = strrchr(argv[0], '/');
 	progname = ((progname == NULL) ? argv[0] : progname+1);
 	signal(SIGINT, ipmi_catch_sigint);
 
-	do
+	while ((argflag = getopt(argc, (char **)argv, OPTION_STRING)) != -1)
 	{
-		if (arge != NULL) {
-			voptind = optind;
-			cnt = argecnt;
-			arg = arge;
-		}
-		while ((argflag = getopt(cnt, (char **)arg, OPTION_STRING)) != -1)
-		{
 		switch (argflag) {
 		case 'I':
 			if (intfname) {
@@ -750,43 +735,12 @@ ipmi_main(int argc, char ** argv,
 			ipmi_option_usage(progname, cmdlist, intflist);
 			goto out_free;
 		}
-		}
-
-		/* support additional arguments from environment */
-		if (argecnt == 0) {
-			argestr = getenv("IPMITOOL_OPTS");
-			if (argestr == NULL) {
-				argestr = getenv("IPMI_OPTS");
-			}
-			if (argestr != NULL) {
-				argecnt = 0;
-				arge = malloc(sizeof(char*)*(argecnt+2));
-				arge[argecnt++] = "ipmitool-env";
-				arge[argecnt] = NULL;
-				while ((arge[argecnt] = strsep(&argestr, " ")) != NULL) {
-					narge = realloc(arge, sizeof(char*)*(argecnt+2));
-					if (narge == NULL) {
-						free(arge);
-						lprintf(LOG_ERR, "Problem while resizing options from environment.\n");
-						rc = -1;
-						goto out_free;
-					} else {
-						arge = narge;
-					}
-					arge[++argecnt] = NULL;
-				}
-
-			}
-		}
-	} while ((argecnt>1) && (arg != arge));
-		if (NULL != arge) {
-		optind = voptind;
-
 	}
 
+
 	/* check for command before doing anything */
 	if (argc-optind > 0 &&
-			strncmp(argv[optind], "help", 4) == 0) {
+		strncmp(argv[optind], "help", 4) == 0) {
 		ipmi_cmd_print(cmdlist);
 		rc = 0;
 		goto out_free;
