diff -up ipmitool-1.8.11/doc/ipmitool.1.k-option ipmitool-1.8.11/doc/ipmitool.1
--- ipmitool-1.8.11/doc/ipmitool.1.k-option	2009-02-25 21:38:52.000000000 +0100
+++ ipmitool-1.8.11/doc/ipmitool.1	2009-02-26 13:27:00.000000000 +0100
@@ -26,7 +26,7 @@ ipmitool [\fB\-c\fR|\fB\-h\fR|\fB\-v\fR|
          [\fB\-o\fR <\fIoemtype\fP>]
          [\fB\-O\fR <\fIsel oem\fP>]
          [\fB\-C\fR <\fIciphersuite\fP>]
-         [\fB\-K\fR|\fB\-k\fR <\fIkg_key\fP>]
+         [\fB\-Y\fR|[\fB\-K\fR|\fB\-k\fR <\fIkg_key\fP>]
          [\fB\-y\fR <\fIhex_kg_key\fP>]
          [\fB\-e\fR <\fIesc_char\fP>]
          <\fIcommand\fP>
@@ -62,6 +62,9 @@ This is not available with all commands.
 Use supplied character for SOL session escape character.  The default
 is to use \fI~\fP but this can conflict with ssh sessions.
 .TP
+\fB\-K\fR
+Read Kg key from IPMI_KGKEY environment variable.
+.TP
 \fB\-k\fR <\fIkey\fP>
 Use supplied Kg key for IPMIv2 authentication.  The default is not to
 use any Kg key.
@@ -72,6 +75,9 @@ hexadecimal format and can be used to sp
 characters. E.g. '-k PASSWORD' and '-y 50415353574F5244' are
 equivalent.
 The default is not to use any Kg key.
+.TP
+\fB\-Y\fR
+Prompt for the Kg key for IPMIv2 authentication.
 .TP 
 \fB\-C\fR <\fIciphersuite\fP>
 The remote server authentication, integrity, and encryption algorithms
diff -up ipmitool-1.8.11/lib/ipmi_main.c.k-option ipmitool-1.8.11/lib/ipmi_main.c
--- ipmitool-1.8.11/lib/ipmi_main.c.k-option	2009-02-25 21:38:52.000000000 +0100
+++ ipmitool-1.8.11/lib/ipmi_main.c	2009-02-26 13:29:08.000000000 +0100
@@ -73,7 +73,7 @@
 #endif
 
 #ifdef ENABLE_ALL_OPTIONS
-# define OPTION_STRING	"I:hVvcgsEKao:H:d:P:f:U:p:C:L:A:t:T:m:S:l:b:B:e:k:y:O:"
+# define OPTION_STRING	"I:hVvcgsEKYao:H:d:P:f:U:p:C:L:A:t:T:m:S:l:b:B:e:k:y:O:"
 #else
 # define OPTION_STRING	"I:hVvcH:f:U:p:d:S:"
 #endif
@@ -227,6 +227,7 @@ ipmi_option_usage(const char * progname,
 	lprintf(LOG_NOTICE, "       -S sdr         Use local file for remote SDR cache");
 #ifdef ENABLE_ALL_OPTIONS
 	lprintf(LOG_NOTICE, "       -a             Prompt for remote password");
+	lprintf(LOG_NOTICE, "       -Y             Prompt for the Kg key for IPMIv2 authentication");
 	lprintf(LOG_NOTICE, "       -e char        Set SOL escape character");
 	lprintf(LOG_NOTICE, "       -C ciphersuite Cipher suite to be used by lanplus interface");
 	lprintf(LOG_NOTICE, "       -k key         Use Kg key for IPMIv2 authentication");
@@ -470,6 +471,22 @@ ipmi_main(int argc, char ** argv,
 				goto out_free;
 			}
 			break;
+		case 'Y':
+#ifdef HAVE_GETPASSPHRASE
+			tmp = getpassphrase("Key: ");
+#else
+			tmp = getpass("Key: ");
+#endif
+			if (tmp != NULL) {
+				if (kgkey)
+					free(kgkey);
+				kgkey = strdup(tmp);
+				if (kgkey == NULL) {
+					lprintf(LOG_ERR, "%s: malloc failure", progname);
+					goto out_free;
+				}
+			}
+			break;
 		case 'U':
 			username = strdup(optarg);
 			if (username == NULL) {
