diff -up ./include/ipmitool/ipmi_sdr.h.undername ./include/ipmitool/ipmi_sdr.h
--- ./include/ipmitool/ipmi_sdr.h.undername	2015-03-04 17:55:53.063811747 +0100
+++ ./include/ipmitool/ipmi_sdr.h	2015-03-04 17:56:05.570842326 +0100
@@ -821,7 +821,7 @@ static const char *sensor_type_desc[] __
 	    "Version Change", "FRU State" };
 
 struct sensor_reading {
-	char		s_id[17];		/* name of the sensor */
+	char		s_id[33];		/* name of the sensor */
 	struct sdr_record_full_sensor    *full;
 	struct sdr_record_compact_sensor *compact;
 	uint8_t		s_reading_valid;	/* read value valididity */
